llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp

```diff
@@ -1623,10 +1623,12 @@ void PGOUseFunc::setBranchWeights() {
       continue;
 
     // We have a non-zero Branch BB.
-    unsigned Size = BBCountInfo.OutEdges.size();
-    SmallVector<uint64_t, 2> EdgeCounts(Size, 0);
+    unsigned OutEdgesCount = BBCountInfo.OutEdges.size();
+    unsigned SuccessorCount = BB.getTerminator()->getNumSuccessors();
+
+    SmallVector<uint64_t, 2> EdgeCounts(SuccessorCount, 0);
```

Comment:
Can you explain why coroutines would produce a BB with a successor with an ID outside BBCountInfo.OutEdges?

From there we can figure out a small test case, too.

Reply:
I'm not an LLVM developer. I don't even know what a successor is in this context.  I observed a crash, and following the source I saw the EdgeCounts vector was sized differently than the range of the index. So I changed the vector size to match the range of the index.

Since I don't observe the crash in llvm main, it could be that it changed so that the range of the index now matches the size of the vector. It could also be (and I think this is likely) that the pgo machinery changed enough that the 18.1.6 based profile no longer triggers the bug. I don't have any way to tell. I believe the change improves the code either way (though only cosmetically in one case).

I'm willing to follow guidance in order to generate a test case, but my knowledge here is almost nil.

Reply:
(sorry for the delay, somehow I missed / got no notifications)

If the source you're observing the crash in is something you share (IIUC it is - earlier paste), could you do this, while in a debugger (assuming the unchanged, buggy code):

- get to the place where you observe the crash
- `p BB.getParent()->dump()`
- also `p BB.dump()`
- also `p Size` and `p s` for the failing `s`

Can you share those in a gist (less inlined stuff in the comments, easier to parse)? (Thanks!)

The first dumps the IR (intermediate representation) of the function that causes the problem. The latter, just the [basic block](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/BasicBlock.h#L47). A function's instructions are organized in a [control flow graph (CFG)](https://en.wikipedia.org/wiki/Control-flow_graph) where the nodes are basic blocks. The CFG is directed, and "successor" of a node X is a node Y for which there is an edge X->Y.

Reply:
https://gist.github.com/avikivity/ee2e860be4dd5179b9f8ee70a2e9f487

Reply:
oh - what's `SuccNum`?

Reply:
```
(gdb) p SuccNum
$5 = 2
```

I can provide a docker image that will drop you into gdb at the crash point, if you want.

Reply:
Depends - if you are interested in digging more into llvm, happy to help along. If not, rather than a docker, you can just give me the module: add `-mllvm -print-before=pgo-instr-use` to your compiler invocation, and paste that output to a gist.

Reply:
This is probably too deep for me to learn much. I don't mind following instructions to deliver new data, I'm just concerned that with multi-hour round-trip times, this can take several weeks to resolve.

Anyway here's the -print-before=pgo-instr-use output: https://gist.github.com/avikivity/f47fa3765bd49e05524578d62c118cfd

Reply:
Ah, I didn't know you were able to run `-print-before=pgo-instr-use`!  I now able to reproduce and I think I can get a reasonable test case.

Reply:
Can you add a comment as to why we need to dimension EdgeCounts like that? (basically "`Removed` edges don't appear in `OutEdges`). Also I'd add an `assert(OutEdgesCount <= SuccessorCount)

otherwise LGTM, but wait also for @ellishg and/or @xur-llvm

Reply:
Done